LastPass password fill implementation based off qute-pass code

welps opened the pull request:
Hi, 

I started using qutebrowser a week ago and saw this wasn't available yet (#3909). This script is mostly existing work from the qute-pass implementation by @cryzed with some few changes for the LastPass CLI differences. 

Here's an [example of it being used](https://i.imgur.com/zA61NrF.gifv) with a lower-quality gif version below:
![Video of script](https://i.imgur.com/zA61NrF.gif)

Just some notes: 
 - `lpass show` allows for a regex search on the _name of the LastPass entry_ and can return all matching entries in json so the flow is quite straightforward.
     - When there are multiple entries, each entry is formatted as LastPass ID, name, url, and username for lack of a better idea
 - As far as I can tell LastPass CLI doesn't have any way to search against an entry's url which makes more sense than searching an entry's name, which can be provided by the user. If this isn't the case, this script should be updated to use that method.
 - `lpass` will prompt for the master password again after an hour, you can set the environment variable `LPASS_AGENT_TIMEOUT` or add it to your `$HOME/.lpass` to the number of seconds you prefer or 0. See `man lpass` if you need help.
 - I removed the username/password matching that qute-pass had since it's not necessary

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/3929)
<!-- Reviewable:end -->

[edit by @jgkamat to link issues: closes #3909]

jgkamat commented on misc/userscripts/qute-lastpass:
```diff
@@ -0,0 +1,159 @@
+"websites/github.com".  The login information is inserted by emulating key events using qutebrowser's fake-key command in this manner:
+[USERNAME]<Tab>[PASSWORD], which is compatible with almost all login forms."""
+
+EPILOG = """Dependencies: tldextract (Python 3 module), LastPass CLI.
```
There should be a note saying that this only supports `lpass` version `1.3` and newer, as that's when `-j` was added

Also, there should be a note telling people to `lpass login <email>` somewhere, for those new to `lpass`.

jgkamat commented on misc/userscripts/qute-lastpass:
```diff
@@ -0,0 +1,159 @@
+
+USAGE = """The domain of the site has to be in the name of the LastPass entry, for example: "github.com/cryzed" or
+"websites/github.com".  The login information is inserted by emulating key events using qutebrowser's fake-key command in this manner:
+[USERNAME]<Tab>[PASSWORD], which is compatible with almost all login forms."""
```
It would help people who haven't read the docs to have an example invocation here, for example:

`:spawn --userscript qute-lastpass`

jgkamat commented on misc/userscripts/qute-lastpass:
```diff
@@ -0,0 +1,159 @@
+    # the registered domain name and finally: the IPv4 address if that's what
+    # the URL represents
+    candidates = []
+    for target in filter(None, [extract_result.fqdn, extract_result.registered_domain, extract_result.ipv4]):
```
I'm not sure if `tldextract` supports this, but it might be useful to get the result without the tld as well. For example:

`github.com` -> `github`.

I think this might help catch some more results, since we can't search the url directly...

jgkamat commented on misc/userscripts/qute-lastpass:
```diff
@@ -0,0 +1,159 @@
+    # the URL represents
+    candidates = []
+    for target in filter(None, [extract_result.fqdn, extract_result.registered_domain, extract_result.ipv4]):
+        target_candidates = json.loads(pass_(target, arguments.io_encoding))
```
This line causes an error for me if there are no results found:

```
Process stderr:
Error: Could not find specified account(s).
Traceback (most recent call last):
  File "/home/jay/Downloads/qute-lastpass", line 159, in <module>
    sys.exit(main(arguments))
  File "/home/jay/Downloads/qute-lastpass", line 111, in main
    target_candidates = json.loads(pass_(target, arguments.io_encoding))
  File "/usr/lib64/python3.5/json/__init__.py", line 319, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python3.5/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python3.5/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

We probably should abort immediately if the return value from the `lpass` call is non-zero (lpass seems to be properly handling that).

welps commented:
@jgkamat Requested changes were made.

- Added additional documentation for usage 
- If LastPass CLI returns an error, print out domain combinations we've tried with their respective error
- Added subdomain + domain, domain only (both without tld) as possible combinations to try.

welps commented on misc/userscripts/qute-lastpass:
```diff
@@ -0,0 +1,172 @@
+    # the registered domain name and finally: the IPv4 address if that's what
+    # the URL represents
+    candidates = []
+    for target in filter(None, {extract_result.fqdn, extract_result.registered_domain, extract_result.subdomain + extract_result.domain, extract_result.domain, extract_result.ipv4}):
```
This is fair as I did not account for merge_candidates. 

Regarding `filter`, are you asking in terms of style or why it's being used? The latter is because several candidates may be an empty string, the former is because this was chosen by the author of the qute-pass implementation.

jgkamat replied:
Oh! I didn't know that `filter(None)` did that, I thought it was just a nop. This looks good :)

welps commented:
@jgkamat Changes made

The-Compiler commented:
Seems fine to me as well! Thanks for the patience and the pull request. :tada:

welps commented:
@The-Compiler Thanks for merging and for making a kickass browser. I was able to abandon using a mouse completely in my day-to-day development workflow and it has been huge efficiency-wise.

vlurgio commented:
So I saw this was merged but how exactly do we go about using it?

welps commented:
@vlurgio 

Run `spawn --userscript qute-lastpass` within qutebrowser. You can bind this command to a key from your local `config.py` for ease of use.
 
See the comments in [qute-lastpass](https://github.com/qutebrowser/qutebrowser/blob/9a85796ac30ef33218dd7dee9db6a3c28364f668/misc/userscripts/qute-lastpass) for additional options.

vlurgio commented:
Yeah that was what I thought although it seems my install (from brew) didn't come with this user-script for some reason

The-Compiler commented:
@vlurgio You can download the userscript from the repo, put it somewhere, and then run `:spawn --userscript` with the full path to it.
